apps/dotcom/sync-worker/src/TLPostgresReplicator.ts

```diff
@@ -120,29 +140,75 @@ export class TLPostgresReplicator extends DurableObject<Environment> {
 
 	private log
 
+	private readonly replicationService
+	private readonly slotName
+	private readonly wal2jsonPlugin = new Wal2JsonPlugin({})
+
 	private readonly db: Kysely<DB>
 	constructor(ctx: DurableObjectState, env: Environment) {
 		super(ctx, env)
 		this.measure = env.MEASURE
 		this.sentry = createSentry(ctx, env)
 		this.sqlite = this.ctx.storage.sql
+		this.state = {
+			type: 'init',
+			promise: promiseWithResolve(),
+		}
 
-		this.ctx.blockConcurrencyWhile(async () =>
-			this._migrate().catch((e) => {
-				this.captureException(e)
-				throw e
-			})
-		)
+		const slotNameMaxLength = 63 // max postgres identifier length
+		const slotNamePrefix = 'tlpr_' // pick something short so we can get more of the durable object id
+		const durableObjectId = this.ctx.id.toString()
+		this.slotName =
+			slotNamePrefix + durableObjectId.slice(0, slotNameMaxLength - slotNamePrefix.length)
 
-		// debug logging in preview envs by default
 		this.log = new Logger(env, 'TLPostgresReplicator', this.sentry)
 		this.db = createPostgresConnectionPool(env, 'TLPostgresReplicator', 100)
 
+		this.replicationService = new LogicalReplicationService(
+			/**
+			 * node-postgres Client options for connection
+			 * https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/pg/index.d.ts#L16
+			 */
+			{
+				database: 'postgres',
+				connectionString: env.BOTCOM_POSTGRES_CONNECTION_STRING,
+				application_name: this.slotName,
+			},
+			/**
+			 * Logical replication service config
+			 * https://github.com/kibae/pg-logical-replication/blob/main/src/logical-replication-service.ts#L9
+			 */
+			{
+				acknowledge: {
+					auto: false,
+					timeoutSeconds: 10,
+				},
+			}
+		)
+
 		this.alarm()
-		this.reboot('constructor', false).catch((e) => {
-			this.captureException(e)
-			this.__test__panic()
-		})
+		this.ctx
+			.blockConcurrencyWhile(async () => {
+				await this._migrate().catch((e) => {
+					this.captureException(e)
+					throw e
+				})
+				// if the slot name changed, we set the lsn to null, which will trigger a mass user DO reboot
+				if (this.sqlite.exec('select slotName from meta').one().slotName !== this.slotName) {
+					this.sqlite.exec('UPDATE meta SET slotName = ?, lsn = null', this.slotName)
+				}
+				await sql`SELECT pg_create_logical_replication_slot(${this.slotName}, 'wal2json') WHERE NOT EXISTS (SELECT 1 FROM pg_replication_slots WHERE slot_name = ${this.slotName})`.execute(
```

Comment:
Is `this.slotName` guaranteed to be always the same (not sure if DO ids are guaranteed to be the same across deploys, reboots,...)?

Reply:
yeah the DO ID is guaranteed to be stable. 

this.slotName would only change if we manually alter the prefix and redeploy, and we'd only need to do that if something goes horribly wrong and we need to force all the active user DOs to reboot from scratch. I don't think we'll ever need to do that.